prov/util/src/util_srx.c

```diff
@@ -0,0 +1,1121 @@
+		           void *optval, size_t *optlen)
+{
+	struct util_srx_ctx *srx =
+		container_of(fid, struct util_srx_ctx, peer_srx.ep_fid.fid);
+
+	if ((level != FI_OPT_ENDPOINT) || (optname != FI_OPT_MIN_MULTI_RECV))
+		return -FI_ENOPROTOOPT;
+
+	*(size_t *)optval = srx->min_multi_recv_size;
+	*optlen = sizeof(size_t);
+
+	return FI_SUCCESS;
+}
+
+static int util_srx_setopt(fid_t fid, int level, int optname,
+		           const void *optval, size_t optlen)
+{
+	struct util_srx_ctx *srx =
+		container_of(fid, struct util_srx_ctx, peer_srx.ep_fid.fid);
+
+	if ((level != FI_OPT_ENDPOINT) || (optname != FI_OPT_MIN_MULTI_RECV))
+		return -FI_ENOPROTOOPT;
+
+	srx->min_multi_recv_size = *(size_t *)optval;
+
+	return FI_SUCCESS;
+}
+
+static struct fi_ops_ep util_srx_ops = {
+	.size = sizeof(struct fi_ops_ep),
+	.cancel = util_srx_cancel,
+	.getopt = util_srx_getopt,
+	.setopt = util_srx_setopt,
+	.tx_ctx = fi_no_tx_ctx,
+	.rx_ctx = fi_no_rx_ctx,
+	.rx_size_left = fi_no_rx_size_left,
+	.tx_size_left = fi_no_tx_size_left,
+};
+
+static void util_rx_entry_init_fn(struct ofi_bufpool_region *region, void *buf)
+{
+	struct util_rx_entry *rx_entry = (struct util_rx_entry *) buf;
+	struct util_srx_ctx *srx = (struct util_srx_ctx *)
+					region->pool->attr.context;
+
+	rx_entry->peer_entry.iov = util_srx_iov(rx_entry);
+	rx_entry->peer_entry.desc = util_srx_desc(srx, rx_entry);
+}
+
+static void util_srx_init_slist(struct ofi_dyn_arr *arr, void *item)
+{
+	slist_init((struct slist *) item);
+}
+
+static void util_srx_init_dlist(struct ofi_dyn_arr *arr, void *item)
+{
+	dlist_init((struct dlist_entry *) item);
+}
+
+int util_ep_srx_context(struct util_domain *domain, size_t rx_size,
+			size_t iov_limit, size_t default_min_mr,
+			struct fid_ep **rx_ep)
+{
+	struct util_srx_ctx *srx;
+	struct ofi_bufpool_attr pool_attr;
+	int ret = FI_SUCCESS;
+
+	srx = calloc(1, sizeof(*srx));
+	if (!srx)
+		return -FI_ENOMEM;
+
+	ofi_mutex_init(&srx->lock);
+	if (domain->threading != FI_THREAD_SAFE) {
+		srx->lock_acquire = ofi_mutex_lock_noop;
+		srx->lock_release = ofi_mutex_unlock_noop;
+	} else {
+		srx->lock_acquire = ofi_mutex_lock_op;
+		srx->lock_release = ofi_mutex_unlock_op;
+	}
+
+	ofi_array_init(&srx->src_unexp_msg_queues, sizeof(struct dlist_entry),
+		       util_srx_init_dlist);
+	ofi_array_init(&srx->src_unexp_tag_queues, sizeof(struct dlist_entry),
+		       util_srx_init_dlist);
+
+	ofi_array_init(&srx->src_recv_queues, sizeof(struct slist),
+		       util_srx_init_slist);
+	ofi_array_init(&srx->src_trecv_queues, sizeof(struct slist),
+		       util_srx_init_slist);
+
+	slist_init(&srx->msg_queue);
+	slist_init(&srx->tag_queue);
+
+	dlist_init(&srx->all_unexp_msg);
+	dlist_init(&srx->all_unexp_tag);
+
+	dlist_init(&srx->progress_msg_queue);
+	dlist_init(&srx->progress_tag_queue);
+	srx->rx_seq_no = 0;
+
+	pool_attr.size = sizeof(struct util_rx_entry) +
+		(sizeof(struct iovec) + sizeof(void *)) * iov_limit;
```

Comment:
Nit: I would add comment here to explain the math of `(sizeof(struct iovec) + sizeof(void *)) * iov_limit`. I understand it is needed to count the size of `iov` and `desc`.

 In the future it may be better to make util_rx_entry have a `data` field so providers can embed additional context inside it, like the util_av_entry today. But it does not have to be part of this PR.